### PR TITLE
Implement Newton DFT solver

### PR DESCRIPTION
**Pull request by prehner**
This PR implements a quickly converging Newton DFT solver that uses a matrix-free linear solver (GMRES) to calculate the linear subsystem. The action fo the Jacobian on the Newton step is calculated efficiently using fast Fourier transforms.

In addition, a solver log is added to every `DFTProfile` that keeps track of the residual and the execution time of the solver. This allows a detailed analysis of the choice of solver and solver parameters for any DFT calculation.

closing #74

### CODE REVIEW

File: feos-dft/src/profile.rs [hunk 1]
```diff
@@ -389,110 +383,71 @@ where
...
+        } else {
+            &rho_projected - density
+        };
+
         // set residual to 0 where external potentials are overwhelming
-        res_rho
-            .iter_mut()
+        res.iter_mut()
             .zip(self.external_potential.iter())
             .for_each(|(r, &p)| {
```
Comment (g-bauer) [inline feos-dft/src/profile.rs:410]: (not a recent change but) would this be more clear with `filter().for_each()` instead of the `if` case?

**Review by g-bauer — CHANGES_REQUESTED** [review]
Two overarching comments:

- the damping coefficient is \nu in the documentation and \beta in the code and in the interfaces which can be confusing when going through the theory section and then work with the code - should we change it to be \nu in the interfaces/code as well (breaking change)?
- nitpick: `dr` vs `\mathrm{d}r` for integrals is different in each document. We should decide which variant to use and use it everywhere.

Awesome PR!

File: docs/theory/dft/solver.md [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+&=\int\sum_\beta\frac{\delta}{\delta\rho_\beta(r')}\left(\rho_\alpha(r)-\rho_\alpha^\mathrm{b}e^{\frac{\beta}{m_\alpha}\left(\hat F_{\rho_\alpha}^\mathrm{b,res}-\hat F_{\rho_\alpha}^\mathrm{res}(r)-V_\alpha^\mathrm{ext}(r)\right)}\prod_{\alpha'}I_{\alpha\alpha'}(r)\right)\Delta\rho_\beta(r')dr'
+\end{align*}$$
+
+The functional derivative can be simplified using $\hat F_{\rho_\alpha\rho_\beta}^\mathrm{res}(r,r')=\frac{\delta \hat F_{\rho_\alpha}^\mathrm{b,res}(r)}{\delta\rho_\beta(r')}=\frac{\delta^2\hat F^\mathrm{b,res}}{\delta\rho_\alpha(r)\delta\rho_\beta(r')}$
+
+$$\begin{align*}
+q_\alpha(r)&=\int\sum_\beta\left(\delta_{\alpha\beta}\delta(r-r')+\left(\frac{\beta}{m_\alpha}\hat F_{\rho_\alpha\rho_\beta}^\mathrm{res}(r,r')-\sum_{\alpha'}\frac{1}{I_{\alpha\alpha'}(r)}\frac{\delta I_{\alpha\alpha'}(r)}{\delta\rho_\beta(r')}\right)\right.\\
+&~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\left.\times\rho_\alpha^\mathrm{b}e^{\frac{\beta}{m_\alpha}\left(\hat F_{\rho_\alpha}^\mathrm{b,res}-\hat F_{\rho_\alpha}^\mathrm{res}(r)-V_\alpha^\mathrm{ext}(r)\right)}\prod_{\alpha'}I_{\alpha\alpha'}(r)\right)\Delta\rho_\beta(r')dr'\\
```
[the remark this answers is not in the document]
Comment (prehner) [reply docs/theory/dft/solver.md:66]: I'm not sure what you mean. Aligning, e.g., at the equal sign is less readable imo. I tried to emulate a `multline` block which for some reason I couldn't get to work.

File: feos-dft/src/profile.rs (hunk 1 above)
Comment (prehner) [reply feos-dft/src/profile.rs:410]: Me gusta!

**Comment by prehner** [comment]
You are right! The notation in the theory guide should follow the code as close as possible. That is the wohle point of it.

So beta is already used of course. alpha is also commonly used when discussing line search, but alpha is used as a segment index. I went with the very explicit `damping_coefficient` instead of a spelled out greek letter. This is indeed a breaking change in Python but I think that is ok.

**Review by g-bauer — APPROVED** [review]

**Pull request merged**